### ISSUE TITLE
Auto lowercase of tag name and attribute name

### ISSUE THREAD

**Issue by songsiqi** [inikulin/parse5#116]
While parsing `<customTag customAttr="a">`, tag name is `customtag` and property name is `customattr`, it's not very friendly. We want `customTag` or `custom-tag`. Is this configurable?

**Comment by RReverser** [comment]
@songsiqi This is how browsers parse HTML as well - otherwise you might get ambiguity (for example, `customAttr` and `customattr` and `CustomAttr` are all treated by browser as a same attribute).

I believe it's not hard to perform `.toLowerCase()` in own `getAttr` helper if all that you want is to call in `getAttr('customAttr')` fashion.

**Comment by inikulin** [comment]
> We want customTag or custom-tag

Can I ask why you need this? AFAIK DOM API is case insensitive for HTML documents. Case matters when you deal with XML (XHTML), but you need XML parser then.

**Comment by songsiqi** [comment]
@inikulin Any XML parser recommended?

**Comment by RReverser** [comment]
@songsiqi If you're in browser, you can go away with `new DOMParser().parseFromString(str, 'text/xml')`, in Node.js you might want to take a look at https://github.com/Leonidas-from-XIV/node-xml2js.

**Comment by domenic** [comment]
Fun fact: there is actually no spec for parsing XML into DOM structures, only for parsing it into something called an "infoset".

**Comment by RReverser** [comment]
> An XML parser, for the purposes of this specification, is a construct that follows the rules given in the XML specification to map a string of bytes or characters into a Document object.

> At the time of writing, no such rules actually exist.

What?!

**Comment by RReverser** [comment]
@domenic Well, looking deeper, it refers to [XML spec](https://www.w3.org/TR/xml/) which has set of rules to convert byte stream into nodes,  and [HTML spec](https://html.spec.whatwg.org/multipage/xhtml.html) has this section:

> This Document must then be populated with DOM nodes that represent the tree structure of the input passed to the parser, as defined by the XML specification, the Namespaces in XML specification, and the WHATWG DOM standard. DOM mutation events must not fire for the operations that the XML parser performs on the Document's tree, but the user agent must act as if elements and attributes were individually appended and set respectively so as to trigger rules in this specification regarding what happens when an element is inserted into a document or has its attributes set, and the WHATWG DOM standard's requirements regarding mutation observers mean that mutation observers are fired (unlike mutation events).

which kind of describes interop, although in non-normative way.

**Comment by inikulin** [comment]
@RReverser But it's still missing actual rules for the mapping. E.g. how procesing instructions should be handled, how to deal with case-sensitivity

**Comment by RReverser** [comment]
@inikulin It's mainly because they are mapped pretty much 1:1 (XML names are kept in generated DOM without any extra transformations - e.g. `new DOMParser().parseFromString('<XML />', 'text/xml').children[0].tagName == "XML"`). But I agree that it should be stated more explicitly.

**Comment by domenic** [comment]
Yeah, I don't think this is really accurate based on reading the XML spec:

> it refers to XML spec which has set of rules to convert byte stream into nodes

There's no actual parser algorithm, just some kind of vague correspondence. The word "node" appears exactly once in the document, BTW.

**Comment by inikulin** [comment]
@songsiqi Can I close issue?

**Comment by Meir017** [comment]
in order to support this you just need to modify `packages\parse5\lib\tokenizer\index.js` and join the cases of `isAsciiUpper(cp)` with the `isAsciiLower(cp)` ones @songsiqi

**Comment by tenadolanter** [comment]
why not provide a config, let user choose  @RReverser 

in my case, i want to parser my .vue file,  some code whould be like  `<Test originType="WORKSPACE">`

use the parser, then i get `<test origintype="WORKSPACE">`, it change my code

**Comment by wooorm** [comment]
This project can’t parse Vue. This idea here would solve one tiny part of your problem but there are still 100s of differences that don’t work. Use a Vue parser.

Locking, this is an 7 year old issue that is not going to happen
